File: api/src/routes/features.ts
import { readFileSync } from "node:fs"
import { Hono } from "hono"
import {
  listFeatures,
  getFeature,
  createFeature,
  updateFeature,
  deleteFeature,
  listFeatureArtifacts,
  reorderFeatures,
  getArtifactContent,
  getArtifactFilePath,
  saveArtifactContent,
  saveArtifactBuffer,
  deleteArtifact,
  isTextArtifact,
} from "../store/features.js"
import type { FeatureStatus, Priority } from "../store/features.js"
import { updateCard } from "../store/roadmap.js"
import {
  CreateFeatureSchema,
  UpdateFeatureSchema,
  ReorderSchema,
  CreateArtifactSchema,
  SaveArtifactContentSchema,
  validateBody,
} from "../lib/validation.js"
import {
  serializeFeature,
  serializePaginatedFeatures,
  serializeFeatureArtifact,
  serializeArtifactContent,
} from "../lib/serialize.js"

const router = new Hono()

// GET /api/features — paginated listing with filters
router.get("/", (c) => {
  const url = new URL(c.req.url)

  const page = Math.max(1, parseInt(url.searchParams.get("page") ?? "1", 10) || 1)
  const limit = Math.max(1, Math.min(100, parseInt(url.searchParams.get("limit") ?? "25", 10) || 25))

  // Filters
  const statusParam = url.searchParams.get("status")
  const status = statusParam
    ? statusParam.split(",").filter((s): s is FeatureStatus =>
        ["draft", "deferred", "ready", "in-progress", "review", "done"].includes(s))
    : undefined

  const priorityParam = url.searchParams.get("priority")
  const priority = priorityParam
    ? priorityParam.split(",").filter((p): p is Priority =>
        ["critical", "high", "medium", "low"].includes(p))
    : undefined

  const assigneeParam = url.searchParams.get("assignee")
  let assignee: string | null | undefined = undefined
  if (assigneeParam !== null) {
    assignee = assigneeParam === "__unassigned__" ? null : assigneeParam
  }

  const pointsMinParam = url.searchParams.get("points_min")
  const pointsMin = pointsMinParam ? Number(pointsMinParam) : undefined

  const pointsMaxParam = url.searchParams.get("points_max")
  const pointsMax = pointsMaxParam ? Number(pointsMaxParam) : undefined

  const tagsParam = url.searchParams.get("tags")
  const tags = tagsParam ? tagsParam.split(",").filter(Boolean) : undefined

  // Sorting
  const sortParam = url.searchParams.get("sort") ?? "order"
  const sort = (["order", "title", "status", "priority", "modified"].includes(sortParam)
    ? sortParam
    : "order") as "order" | "title" | "status" | "priority" | "modified"

  const dirParam = url.searchParams.get("dir") ?? "asc"
  const dir = dirParam === "desc" ? "desc" : "asc"

  const result = listFeatures({
    page,
    limit,
    status,
    priority,
    assignee,
    pointsMin: pointsMin !== undefined && !isNaN(pointsMin) ? pointsMin : undefined,
    pointsMax: pointsMax !== undefined && !isNaN(pointsMax) ? pointsMax : undefined,
    tags,
    sort,
    dir,
  })

  return c.json(serializePaginatedFeatures(result))
})

// PATCH /api/features/reorder — reorder features by slug array
router.patch("/reorder", async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(ReorderSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }
  const offset = parsed.data.offset ?? 0
  reorderFeatures(parsed.data.slugs, offset)
  return c.json({ ok: true })
})

// GET /api/features/:slug
router.get("/:slug", (c) => {
  const slug = c.req.param("slug")
  const feature = getFeature(slug)
  if (!feature) {
    return c.json({ error: "Feature not found", slug }, 404)
  }
  return c.json(serializeFeature(feature))
})

// POST /api/features
router.post("/", async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(CreateFeatureSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  try {
    const feature = createFeature({
      title: parsed.data.title,
      body: parsed.data.body,
      priority: parsed.data.priority,
      roadmapCardId: parsed.data.roadmap_card_id,
    })

    // If created from a roadmap card, link the card back to this feature
    if (parsed.data.roadmap_card_id) {
      updateCard(parsed.data.roadmap_card_id, { featureSlug: feature.slug })
    }

    return c.json(serializeFeature(feature), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to create feature"
    if (message.includes("already exists")) {
      return c.json({ error: message }, 409)
    }
    return c.json({ error: message }, 500)
  }
})

// PUT /api/features/:slug
router.put("/:slug", async (c) => {
  const slug = c.req.param("slug")
  const body = await c.req.json()
  const parsed = validateBody(UpdateFeatureSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  try {
    const feature = updateFeature(slug, {
      title: parsed.data.title,
      status: parsed.data.status,
      priority: parsed.data.priority,
      assignees: parsed.data.assignees,
      points: parsed.data.points,
      tags: parsed.data.tags,
      body: parsed.data.body,
    })

    if (!feature) {
      return c.json({ error: "Feature not found", slug }, 404)
    }

    return c.json(serializeFeature(feature))
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to update feature"
    return c.json({ error: message }, 500)
  }
})

// DELETE /api/features/:slug
router.delete("/:slug", (c) => {
  const slug = c.req.param("slug")
  const deleted = deleteFeature(slug)
  if (!deleted) {
    return c.json({ error: "Feature not found", slug }, 404)
  }
  return c.json({ ok: true, slug })
})

// GET /api/features/:slug/artifacts
router.get("/:slug/artifacts", (c) => {
  const slug = c.req.param("slug")
  const artifacts = listFeatureArtifacts(slug)
  if (artifacts === null) {
    return c.json({ error: "Feature not found", slug }, 404)
  }
  return c.json(artifacts.map(serializeFeatureArtifact))
})

// POST /api/features/:slug/artifacts/upload — upload a file
router.post("/:slug/artifacts/upload", async (c) => {
  const slug = c.req.param("slug")
  const formData = await c.req.formData()
  const file = formData.get("file")

  if (!file || !(file instanceof File)) {
    return c.json({ error: "file is required" }, 400)
  }

  const filename = file.name
  if (!filename || filename === "FEATURE.md") {
    return c.json({ error: "Invalid filename" }, 400)
  }

  try {
    const arrayBuffer = await file.arrayBuffer()
    const buffer = Buffer.from(arrayBuffer)
    const artifact = saveArtifactBuffer(slug, filename, buffer)
    if (!artifact) {
      return c.json({ error: "Feature not found", slug }, 404)
    }
    return c.json(serializeFeatureArtifact(artifact), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to upload file"
    return c.json({ error: message }, 500)
  }
})

// POST /api/features/:slug/artifacts/create — create a new empty text file
router.post("/:slug/artifacts/create", async (c) => {
  const slug = c.req.param("slug")
  const body = await c.req.json()
  const parsed = validateBody(CreateArtifactSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  const filename = parsed.data.filename
  if (filename === "FEATURE.md") {
    return c.json({ error: "Invalid filename" }, 400)
  }

  // Check if file already exists
  const existing = getArtifactFilePath(slug, filename)
  if (existing) {
    return c.json({ error: "File already exists", filename }, 409)
  }

  try {
    const artifact = saveArtifactContent(slug, filename, "")
    if (!artifact) {
      return c.json({ error: "Feature not found", slug }, 404)
    }
    return c.json(serializeFeatureArtifact(artifact), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to create file"
    return c.json({ error: message }, 500)
  }
})

// GET /api/features/:slug/artifacts/:filename — get artifact content
router.get("/:slug/artifacts/:filename", (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const url = new URL(c.req.url)
  const raw = url.searchParams.has("raw")

  // If ?raw or binary file, serve the raw file
  if (raw || !isTextArtifact(filename)) {
    const result = getArtifactFilePath(slug, filename)
    if (!result) {
      return c.json({ error: "Artifact not found", slug, filename }, 404)
    }
    const content = readFileSync(result.filePath)
    return new Response(content, {
      headers: {
        "Content-Type": result.mimeType,
        "Content-Length": String(content.length),
        "Cache-Control": "no-cache",
      },
    })
  }

  // Text file — return as JSON
  const artifact = getArtifactContent(slug, filename)
  if (!artifact) {
    return c.json({ error: "Artifact not found", slug, filename }, 404)
  }
  return c.json(serializeArtifactContent(artifact))
})

// PUT /api/features/:slug/artifacts/:filename — save artifact content
router.put("/:slug/artifacts/:filename", async (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const body = await c.req.json()
  const parsed = validateBody(SaveArtifactContentSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  const artifact = saveArtifactContent(slug, filename, parsed.data.content)
  if (!artifact) {
    return c.json({ error: "Feature not found or invalid filename", slug, filename }, 404)
  }
  return c.json(serializeFeatureArtifact(artifact))
})

// DELETE /api/features/:slug/artifacts/:filename — delete an artifact
router.delete("/:slug/artifacts/:filename", (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const deleted = deleteArtifact(slug, filename)
  if (!deleted) {
    return c.json({ error: "Artifact not found", slug, filename }, 404)
  }
  return c.json({ ok: true, slug, filename })
})

export default router

File: commandline/internal/handler/features.go
package handler

import (
	"encoding/json"
	"net/http"
	"os"

	"github.com/briantol/folio/internal/model"
	"github.com/briantol/folio/internal/store"
	"github.com/go-chi/chi/v5"
)

// FeaturesHandler handles /api/features routes.
type FeaturesHandler struct {
	features *store.FeatureStore
	roadmap  *store.RoadmapStore
}

// NewFeaturesHandler creates a new FeaturesHandler.
func NewFeaturesHandler(f *store.FeatureStore, r *store.RoadmapStore) *FeaturesHandler {
	return &FeaturesHandler{features: f, roadmap: r}
}

// Routes returns a chi router for /api/features.
func (h *FeaturesHandler) Routes() chi.Router {
	r := chi.NewRouter()

	r.Get("/", h.list)
	r.Post("/", h.create)
	r.Patch("/reorder", h.reorder)

	r.Route("/{slug}", func(r chi.Router) {
		r.Get("/", h.get)
		r.Put("/", h.update)
		r.Delete("/", h.delete)

		r.Get("/artifacts", h.listArtifacts)
		r.Post("/artifacts/upload", h.uploadArtifact)
		r.Post("/artifacts/create", h.createArtifact)

		r.Route("/artifacts/{filename}", func(r chi.Router) {
			r.Get("/", h.getArtifact)
			r.Put("/", h.saveArtifact)
			r.Delete("/", h.deleteArtifact)
		})
	})

	return r
}

func (h *FeaturesHandler) list(w http.ResponseWriter, r *http.Request) {
	page := QueryInt(r, "page", 1, 1, 999999)
	limit := QueryInt(r, "limit", 25, 1, 100)

	// Assignee filter
	var assignee *string
	if v := r.URL.Query().Get("assignee"); v != "" {
		if v == "__unassigned__" {
			empty := ""
			assignee = &empty
		} else {
			assignee = &v
		}
	}

	params := model.ListFeaturesParams{
		Page:      page,
		Limit:     limit,
		Status:    QueryCSV(r, "status"),
		Priority:  QueryCSV(r, "priority"),
		Assignee:  assignee,
		PointsMin: QueryFloat(r, "points_min"),
		PointsMax: QueryFloat(r, "points_max"),
		Tags:      QueryCSV(r, "tags"),
		Sort:      r.URL.Query().Get("sort"),
		Dir:       r.URL.Query().Get("dir"),
	}
	if params.Sort == "" {
		params.Sort = "order"
	}
	if params.Dir == "" {
		params.Dir = "asc"
	}

	result := h.features.List(params)
	JSON(w, 200, result)
}

func (h *FeaturesHandler) get(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	feature := h.features.Get(slug)
	if feature == nil {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}
	JSON(w, 200, feature)
}

func (h *FeaturesHandler) create(w http.ResponseWriter, r *http.Request) {
	var input model.CreateFeatureInput
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	if input.Title == "" {
		ErrorJSON(w, 422, "title: Title is required")
		return
	}

	feature, err := h.features.Create(input)
	if err != nil {
		msg := err.Error()
		if os.IsExist(err) || containsString(msg, "already exists") {
			ErrorJSON(w, 409, msg)
			return
		}
		ErrorJSON(w, 500, msg)
		return
	}

	// If created from a roadmap card, link the card back
	if input.RoadmapCardID != nil {
		h.roadmap.UpdateCard(*input.RoadmapCardID, map[string]interface{}{
			"featureSlug": feature.Slug,
		})
	}

	JSON(w, 201, feature)
}

func (h *FeaturesHandler) update(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")

	// Parse with raw JSON to detect which fields are present
	var rawMap map[string]json.RawMessage
	if err := ReadJSON(r, &rawMap); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	var input model.UpdateFeatureInput

	if v, ok := rawMap["title"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			if s == "" {
				ErrorJSON(w, 422, "title: String must contain at least 1 character(s)")
				return
			}
			input.Title = &s
		}
	}
	if v, ok := rawMap["status"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Status = &s
		}
	}
	if v, ok := rawMap["priority"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Priority = &s
		}
	}
	if v, ok := rawMap["assignees"]; ok {
		input.AssigneesSet = true
		json.Unmarshal(v, &input.Assignees)
	}
	if v, ok := rawMap["points"]; ok {
		input.PointsSet = true
		json.Unmarshal(v, &input.Points)
	}
	if v, ok := rawMap["tags"]; ok {
		input.TagsSet = true
		json.Unmarshal(v, &input.Tags)
	}
	if v, ok := rawMap["body"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Body = &s
		}
	}

	feature, err := h.features.Update(slug, input)
	if err != nil {
		ErrorJSON(w, 500, err.Error())
		return
	}
	if feature == nil {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}

	JSON(w, 200, feature)
}

func (h *FeaturesHandler) delete(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	if !h.features.Delete(slug) {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}
	JSON(w, 200, map[string]interface{}{"ok": true, "slug": slug})
}

func (h *FeaturesHandler) reorder(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Slugs  []string `json:"slugs"`
		Offset int      `json:"offset"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}
	if input.Slugs == nil {
		ErrorJSON(w, 422, "slugs: Required")
		return
	}
	h.features.Reorder(input.Slugs, input.Offset)
	JSON(w, 200, map[string]interface{}{"ok": true})
}

func (h *FeaturesHandler) listArtifacts(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	artifacts, ok := h.features.ListArtifacts(slug)
	if !ok {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}
	JSON(w, 200, artifacts)
}

func (h *FeaturesHandler) uploadArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")

	if err := r.ParseMultipartForm(32 << 20); err != nil {
		ErrorJSON(w, 400, "Invalid multipart form")
		return
	}

	file, header, err := r.FormFile("file")
	if err != nil {
		ErrorJSON(w, 400, "file is required")
		return
	}
	defer file.Close()

	filename := header.Filename
	if filename == "" || filename == "FEATURE.md" {
		ErrorJSON(w, 400, "Invalid filename")
		return
	}

	data := make([]byte, header.Size)
	if _, err := file.Read(data); err != nil {
		ErrorJSON(w, 500, "Failed to read file")
		return
	}

	artifact := h.features.SaveArtifactBuffer(slug, filename, data)
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}
	JSON(w, 201, artifact)
}

func (h *FeaturesHandler) createArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	var input struct {
		Filename string `json:"filename"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}
	if input.Filename == "" {
		ErrorJSON(w, 422, "filename: Filename is required")
		return
	}
	if input.Filename == "FEATURE.md" {
		ErrorJSON(w, 400, "Invalid filename")
		return
	}

	// Check if exists
	if _, _, ok := h.features.GetArtifactFilePath(slug, input.Filename); ok {
		JSON(w, 409, map[string]interface{}{"error": "File already exists", "filename": input.Filename})
		return
	}

	artifact := h.features.SaveArtifactContent(slug, input.Filename, "")
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found", "slug": slug})
		return
	}
	JSON(w, 201, artifact)
}

func (h *FeaturesHandler) getArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")
	_, raw := r.URL.Query()["raw"]

	if raw || !store.IsTextArtifact(filename) {
		path, mime, ok := h.features.GetArtifactFilePath(slug, filename)
		if !ok {
			JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
			return
		}
		w.Header().Set("Content-Type", mime)
		w.Header().Set("Cache-Control", "no-cache")
		http.ServeFile(w, r, path)
		return
	}

	content := h.features.GetArtifactContent(slug, filename)
	if content == nil {
		JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, content)
}

func (h *FeaturesHandler) saveArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")

	var input struct {
		Content string `json:"content"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	artifact := h.features.SaveArtifactContent(slug, filename, input.Content)
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Feature not found or invalid filename", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, artifact)
}

func (h *FeaturesHandler) deleteArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")
	if !h.features.DeleteArtifact(slug, filename) {
		JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, map[string]interface{}{"ok": true, "slug": slug, "filename": filename})
}

func containsString(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && len(substr) > 0 && findSubstring(s, substr))
}

func findSubstring(s, substr string) bool {
	for i := 0; i+len(substr) <= len(s); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
